rust/src/state_machine/phases/update.rs

```diff
@@ -133,8 +157,26 @@ impl Handler for PhaseState<Update> {
     /// receive a [`PetError::InvalidMessage`].
     fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), PetError> {
         match req {
-            StateMachineRequest::Update(update_req) => self.handle_update(update_req),
-            _ => Err(PetError::InvalidMessage),
+            StateMachineRequest::Update(update_req) => {
+                metrics!(
+                    self.metrics_tx,
+                    metrics::message::update::increment(
+                        self.coordinator_state.round_id,
+                        Self::NAME
+                    )
+                );
+                self.handle_update(update_req)
+            }
+            _ => {
```

Comment:
This patterns comes back a few times. I wonder if there would be a way to use a `Rejection` trait or to extend the existing `Handler` trait to get rid of the repetition.

Reply:
I added a `reject_request` method to the `Handler` trait

Reply:
It doesn't work the way I think it will 😔

I first tried adding `reject_request` to the `Handler` trait

```rust
pub trait Handler {
    fn reject_request(&mut self) -> Result<(), PetError> {
        metrics!(
            self.shared.io.metrics_tx,
            metrics::message::rejected::increment(self.shared.state.round_id, Self::NAME)
        );
        Err(PetError::InvalidMessage)
    }
}
```

but here I get an error because I cannot access `self.shared` or `Self:NAME` (which makes sense).

then I tried

```rust
impl<S> Handler for PhaseState<S>
where
    Self: Phase,
{
    /// Rejects a request by returning a [`PetError::InvalidMessage`] back to the caller.
    fn reject_request(&mut self) -> Result<(), PetError> {
        metrics!(
            self.shared.io.metrics_tx,
            metrics::message::rejected::increment(self.shared.state.round_id, Self::NAME)
        );
        Err(PetError::InvalidMessage)
    }
}
```
and get:
```
error[E0119]: conflicting implementations of trait `state_machine::phases::Handler` for type `state_machine::phases::PhaseState<state_machine::phases::idle::Idle>`:
   --> src/state_machine/phases/mod.rs:168:1
    |
168 | / impl<S> Handler for PhaseState<S>
169 | | where
170 | |     Self: Phase,
171 | | {
...   |
179 | |     }
180 | | }
    | |_^ conflicting implementation for `state_machine::phases::PhaseState<state_machine::phases::idle::Idle>`
    | 
   ::: src/state_machine/phases/idle.rs:23:1
    |
23  |   impl Handler for PhaseState<Idle> {
    |   --------------------------------- first implementation here
```
then I tried to add a `Rejection` trait

```rust
pub trait Rejection {
    /// Rejects a request by returning a [`PetError`] back to the caller.
    fn reject_request(&mut self) -> Result<(), PetError>;
}

impl<S> Rejection for PhaseState<S>
where
    Self: Phase,
{
    /// Rejects a request by returning a [`PetError::InvalidMessage`] back to the caller.
    fn reject_request(&mut self) -> Result<(), PetError> {
        metrics!(
            self.shared.io.metrics_tx,
            metrics::message::rejected::increment(self.shared.state.round_id, Self::NAME)
        );
        Err(PetError::InvalidMessage)
    }
}


impl Handler for PhaseState<Sum>
where
    Self: Rejection,
{
    fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), PetError> {
        match req {
            StateMachineRequest::Sum(sum_req) => {
...
                self.handle_sum(sum_req)
            }
            _ => self.reject_request(),
        }
    }
}
```

but this also didn't work

```
error[E0275]: overflow evaluating the requirement `state_machine::phases::PhaseState<state_machine::phases::sum::Sum>: state_machine::phases::Rejection`
  --> src/state_machine/phases/sum.rs:34:1
   |
34 | / impl Handler for PhaseState<Sum>
35 | | where
36 | |     Self: Rejection,
37 | | {
...  |
54 | |     }
55 | | }
   | |_^
   |
   = help: see issue #48214
```

So I dicided to add a `reject_request` method to `impl<S> PhaseState<S> where Self: Phase,`

Reply:
I just commented about that. Maybe we don't even need a trait actually and this could just go to the `PhaseState::process_single` impl. (Not sure though since I haven't tried it)